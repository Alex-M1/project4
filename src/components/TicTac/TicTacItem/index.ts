import TicTacItem from './TicTacItem';

export default TicTacItem;
