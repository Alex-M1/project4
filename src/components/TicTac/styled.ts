import { TIC_TAC, VIEW_OPTIONS } from 'constants/constants';
import styled from 'styled-components';

export const StTicTacContainer = styled.div`
  display: flex;
  height: ${window.innerHeight - VIEW_OPTIONS.HEADER_HEIGHT}px;
  flex-direction: column;
  justify-content : center;
  align-items: center;
`;

export const StTicTacField = styled.div`
  width: ${TIC_TAC.FIELD_SIZE}px;
  height: ${TIC_TAC.FIELD_SIZE}px;
  border: 1px solid black;
  display: flex;
  flex-wrap: wrap;
`;

export const StTurnText = styled.div`
  margin: 0 0 15px;
  font-size: 30px;
  font-weight: 700;
`;
