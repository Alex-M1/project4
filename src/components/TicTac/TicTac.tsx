import React, {useEffect, useMemo, useState} from 'react';
import { IPlayWithBot, IPlayWithOpponent } from 'store/ticTac/types';
import { v4 as uuidv4 } from 'uuid';
import { GAME_SETTINGS, LOCAL_STORAGE as LS } from 'constants/constants';
import { useTranslation } from 'react-i18next';
import { IMatch } from '../_common_/types/constantsTypes';
import { StTicTacContainer, StTicTacField, StTurnText } from './styled';
import TicTacItem from './TicTacItem';

interface IProps {
  match: IMatch,
  squares: string[],
  isGameEnd: boolean,
  createRoomChanel: (payload: any) => void,
  stepWithBot: (payload: IPlayWithBot) => void,
  stepWithOpponent: (payload: IPlayWithOpponent) => void,
  setTurn: (payload: boolean) => void,
  myOpponentGame: any;
  isMyTurn: boolean;
}

const TicTac: React.FC<IProps> = ({
  match,
  squares,
  isGameEnd,
  stepWithBot,
  stepWithOpponent,
  isMyTurn,
  createRoomChanel,
  myOpponentGame,
}) => {
  const [turn, setTurn] = useState(isMyTurn);
  const data = JSON.parse(localStorage.getItem(LS.gameOptions));
  useEffect(() => {
    setTurn(true);
  }, []);
  useEffect(() => {
    if (!myOpponentGame.id) {
      createRoomChanel({
        roomId: match.params.id,
        myOpponentGame,
      });
    }
  }, [myOpponentGame]);
  useEffect(() => {
    setTurn(isMyTurn);
  }, [isMyTurn]);
  const { t } = useTranslation();
  const stepHandler = (square: number) => {
    if (data && data.playWith === GAME_SETTINGS.user) {
      stepWithOpponent({ square, id: match.params.id });
    } else {
      stepWithBot({ square, id: match.params.id });
    }
  };

  const renderSquares = useMemo(() => {
    return squares.map((el, i) => (
      <TicTacItem
        key={uuidv4()}
        square={i}
        content={el}
        onClick={(square) => stepHandler(square)}
        isGameEnd={isGameEnd}
      />
    ));
  }, [squares]);
  return (
    <StTicTacContainer>
      {
        data && data.playWith === GAME_SETTINGS.user && (
          <StTurnText>{ turn ? t('your_turn') : t('opponent_turn') }</StTurnText>
        )
      }
      <StTicTacField >
        {renderSquares}
      </StTicTacField>
    </StTicTacContainer>
  );
};

export default TicTac;
