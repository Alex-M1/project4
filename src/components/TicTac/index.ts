import { AppStateType } from 'store/rootReducer';
import { connect } from 'react-redux';
import { doStep, stepWithBot, createRoomChanel, stepWithOpponent, setTurn } from 'store/ticTac/actions';
import { getIsGame, getIsMyTurn, getMyOpponentGame, getSquares } from 'store/ticTac/selectors';
import { getRoomList } from 'store/room/selectors';
import TicTac from './TicTac';

const mapStateToProps = (state: AppStateType) => ({
  squares: getSquares(state),
  isGameEnd: getIsGame(state),
  rooms: getRoomList(state),
  myOpponentGame: getMyOpponentGame(state),
  isMyTurn: getIsMyTurn(state),
});

const mapDispatchToProps = {
  doStep,
  stepWithBot,
  stepWithOpponent,
  setTurn,
  createRoomChanel,
};

export default connect(mapStateToProps, mapDispatchToProps)(TicTac);
