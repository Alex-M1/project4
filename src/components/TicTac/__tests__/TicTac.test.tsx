import { mount, shallow } from 'enzyme';
import React from 'react';
import TicTac from '../TicTac';

jest.mock('react-i18next', () => ({
  useTranslation: jest.fn().mockImplementation(() => ({
    t: jest.fn(),
  })),
}),
);

describe('TicTac', () => {
  let props;
  beforeEach(() => {
    props = {
      match: { params: { id: 'ad' } },
      squares: ['ad', 'dasda'],
      isGameEnd: false,
      createRoomChanel: jest.fn(),
      stepWithBot: jest.fn(),
    };
  });

  it('Should match snapshot', () => {
    const component = shallow(<TicTac {...props} />);
    expect(component.html()).toMatchSnapshot();
  });
  it('should render StTicTacContainer', () => {
    const component = mount(<TicTac {...props} />);
    expect(component.find('styled__StTicTacContainer')).toHaveLength(1);
  });
  it('should render StTicTacField', () => {
    const component = mount(<TicTac {...props} />);
    expect(component.find('styled__StTicTacField')).toHaveLength(1);
  });
  it('should render TicTacItem', () => {
    const component = mount(<TicTac {...props} />);
    expect(component.find('TicTacItem')).toHaveLength(2);
  });
  it('should render TicTac useEffect', () => {
    mount(<TicTac {...props} />);
    expect(props.createRoomChanel).toHaveBeenCalled();
  });
  it('should render TicTacItem click', () => {
    const component = mount(<TicTac {...props} />);
    component.find('TicTacItem').at(0).props().onClick();
    expect(props.stepWithBot).toHaveBeenCalled();
  });
});
