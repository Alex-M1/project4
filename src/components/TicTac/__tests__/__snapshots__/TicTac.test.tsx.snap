// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTac Should match snapshot 1`] = `"<div class=\\"styled__StTicTacContainer-cgfTf hMsSZf\\"><div class=\\"styled__StTurnText-ctlytA hpdmR\\"></div><div class=\\"styled__StTicTacField-cIDziS fthusY\\"><div class=\\"styled__StTicTacItem-erJWA-d cemGHT\\">s</div><div class=\\"styled__StTicTacItem-erJWA-d cemGHT\\">s</div></div></div>"`;
